### PR TITLE
Better diagnostics for bounce notifications

### PR DESCRIPTION
**Pull request by rtyley**
This is a fair bit of refactor/cleanup for Ophan Housekeeper - it was precipitated by my trying to debug the source of a raised level of bounce messages, spending quite some time before I realised it wasn't Ophan! It was [Data Technology's instance of Airflow](https://groups.google.com/a/guardian.co.uk/d/msg/ophan.dev/Ckk1wcDjksA/0ZRwGfcSAgAJ)... but the emails that come from [the AWS SNS topic](https://eu-west-1.console.aws.amazon.com/sns/v3/home?region=eu-west-1#/topic/arn:aws:sns:eu-west-1:021353022223:ses-email-bounce-notifications-for-housekeeper) are so unreadable it was difficult to see that.

The new logging messages should be graphable in the Mighty ELK, and from the mail 'source' (the email address of the sender) it should be easy to determine which project is responsible.

I've done a few other things:

* Switched to play-json case-class-based parsing, as we had an increasing number of fields we needed to read to give good diagnostics.
* Slightly more sophisticated use of Scanamo to give some more concise code. Also, modelled fewer fields from the `ophan-alerts` table as we don't need the others.
* Additional tests to ensure we correctly parse the different kind of messages we're going to get
* Attempted to get JSON-based logging going, hope it does... https://github.com/guardian/cloudwatch-logs-management#scala-lambdas

See also https://github.com/guardian/ophan-data-lake/pull/4486 for the action to update the email addresses Airflow uses.

### CODE REVIEW

**Comment by rtyley** [comment]
I don't see the entry we were expecting in https://elasticsearch-6.ophan.co.uk/_aliases ...

File: src/main/scala/housekeeper/Lambda.scala [hunk 1]
```diff
@@ -2,54 +2,34 @@ package housekeeper
...
-object Lambda {
+  def handleBounce(bounceNotification: BounceNotification): Unit = {
+    val bounce = bounceNotification.bounce
 
-  def removeSpeechmarks(email: JsValue): String = email.toString.replaceAll("\"", "")
+    logger.info(Json.obj(
+      "bounce.type" -> bounce.bounceType,
+      "bounce.permanent" -> bounce.isPermanent,
+      "bounce.mail.source" -> bounceNotification.mail.source
+    ))
```
[the remark this answers is not in the document]
Comment (rtyley) [reply src/main/scala/housekeeper/Lambda.scala]: Although in the past on non-Lambda projects I've been able to use `net.logstash.logback.marker.LogstashMarker` construct a JSON-object like structure, it's difficult to me to see how to use 'Markers' as a JSON-structured object in either `log4j` or `slf4j`.

From https://logging.apache.org/log4j/2.x/manual/markers.html :

> Some important rules about Markers must be considered when using them.
> 
> 1. Markers must be unique. They are permanently registered by name so care should be taken to insure that Markers used in your application are distinct from those in the application's dependencies, unless that is what is desired.
> 2. Parent Markers can be added or removed dynamically. However, this is fairly expensive to do. Instead, it is recommended that the parents be identified when obtaining the Marker the first time as shown in the examples above. Specifically, the set method replaces all the markers in a single operation while add and remove act on only a single Marker at a time.
> 3. Evaluating Markers with multiple ancestors is much more expensive than Markers with no parents. For example, in one set of tests to evaluate whether a Marker matched its grandparent took 3 times longer than evaluating the Marker itself. Even then though, evaluating Markers is inexpensive compared to resolving the callers class name or line number.

https://www.slf4j.org/apidocs/org/slf4j/helpers/BasicMarker.html

Log4J markers seem just too basic to store a key-value map?

File: src/main/scala/housekeeper/Dynamo.scala [hunk 2]
```diff
@@ -1,41 +1,35 @@
...
+  private val ophanAlertsTable = Table[OphanAlert]("ophan-alerts")
+
+  def deleteAllAlertsForEmailAddress(email: String): Unit = {
+    Scanamo.exec(dyClient)(for {
+      alertsForEmail <- ophanAlertsTable.query('email -> email)
+      _ = logger.info(Json.obj(
+        "alertsToDelete.emailAddress" -> email,
+        "alertsToDelete.found.count" -> alertsForEmail.size,
+      ))
+      _ <- alertsForEmail.flatMap(_.toOption).traverse(alert => ophanAlertsTable.delete(alert.primaryKey))
```
Comment (TBonnin) [inline src/main/scala/housekeeper/Dynamo.scala]: nitpicking: `.collect { case Right(a) => a}` is a bit more descriptive than `.flatMap(_.toOption)`: one doesn't need to know what `Either.toOption` actually does

File: src/main/scala/housekeeper/Lambda.scala [hunk 3]
```diff
@@ -2,54 +2,36 @@ package housekeeper
...
+  def handleBounce(bounceNotification: BounceNotification): Unit = {
+    val bounce = bounceNotification.bounce
 
-  def removeSpeechmarks(email: JsValue): String = email.toString.replaceAll("\"", "")
+    logger.info(Map(
+      "bounce.type" -> bounce.bounceType,
+      "bounce.permanent" -> bounce.isPermanent,
+      "bounce.mail.source" -> bounceNotification.mail.source,
+      "bounce.bouncedEmailAddresses" -> bounce.bouncedEmailAddresses
+    ), "Woah")
```
Comment (TBonnin) [inline src/main/scala/housekeeper/Lambda.scala]: Woah?

**Pull request merged**